app/scripts/controllers/threebox.js

```diff
@@ -0,0 +1,114 @@
+const ethUtil = require('ethereumjs-util')
+const ObservableStore = require('obs-store')
+const Box = require('3box/dist/3box.min')
+const log = require('loglevel')
+
+class ThreeBoxController {
+  constructor (opts = {}) {
+    const { preferencesController, keyringController, addressBookController, provider } = opts
+
+    this.preferencesController = preferencesController
+    this.addressBookController = addressBookController
+    this.keyringController = keyringController
+    this.provider = provider
+
+    const initState = {
+      threeBoxAddress: null,
+      threeboxSyncing: false,
+      ...opts.initState,
+      syncDone3Box: false,
+    }
+    this.store = new ObservableStore(initState)
+
+    this.init()
+  }
+
+  async init () {
+    const accounts = await this.keyringController.getAccounts()
+    this.address = accounts[0]
+    if (this.address && !(this.box && this.store.getState().syncDone3Box)) {
+      await this.new3Box(this.address)
+    }
+  }
+
+  async _update3Box ({ type }, newState) {
+    const threeBoxSyncing = this.getThreeBoxSyncingState()
+    if (threeBoxSyncing) {
+      await this.box.private.set(type, JSON.stringify(newState))
+    }
+  }
+
+  async new3Box (address) {
+    const threeBoxSyncing = this.store.getState().threeboxSyncing
+
+    if (threeBoxSyncing) {
+      this.store.updateState({ syncDone3Box: false })
+      this.address = address
+
+      try {
+        const data = ethUtil.bufferToHex(Buffer.from('This app wants to view and update your 3Box profile.', 'utf8'))
+        const contentSignature = await this.keyringController.signPersonalMessage({
+          data,
+          from: address,
+        })
+        this.box = await Box.openBox(address, this.provider, {contentSignature})
+        this.box.onSyncDone(() => {
+          log.debug('3Box onSyncDone')
+          this._restoreFrom3Box()
+          this.store.updateState({
+            syncDone3Box: true,
+            threeBoxAddress: address,
+          })
+        })
+      } catch (e) {
+        console.error(e)
+        throw e
+      }
+    }
+  }
+
+  async _restoreFrom3Box () {
+    const threeBoxSyncing = this.getThreeBoxSyncingState()
+    if (threeBoxSyncing) {
+      const backedUpPreferences = await this.box.private.get('preferences')
+      backedUpPreferences && this.preferencesController.store.updateState(JSON.parse(backedUpPreferences))
+      const backedUpAddressBook = await this.box.private.get('addressBook')
+      backedUpAddressBook && this.addressBookController.update(JSON.parse(backedUpAddressBook), true)
+      this._registerUpdates()
+      this.store.updateState({ syncDone3Box: true, threeBoxAddress: this.address })
+    }
+  }
+
+  setThreeBoxSyncing (newThreeboxSyncingState) {
+    const currentState = this.store.getState()
+    this.store.updateState({
+      ...currentState,
```

Comment:
this should be unnecessary for a store.updateState function just passed the single key should only update that key

Reply:
not to mention this will overwrite the new changes

the `...` operator seems to execute after so for example:
```
> let a = {y:1, b:2}
undefined 
> let d = {...a,  y:3}
undefined
> c
{ y: 1, b: 2 }
```